docs/logging.md

```diff
@@ -0,0 +1,93 @@
+
+```
+/var/run/external-logging.sock
+```
+
+The `init`/`service` process will look for this socket and redirect the
+`stdout` and `stderr` of both `onboot` and `services` to `memlogd`.
+
+## memlogd: an in-memory circular buffer
+
+The `memlogd` daemon reads the logs from the `onboot` and `services` containers
+and stores them together with a timestamp and the name of the originating
+container in a circular buffer in memory.
+
+The contents of the circular buffer can be read over the Unix domain socket
+```
+/var/run/memlogq.sock
+```
+
+The circular buffer has a fixed size (overridden by the command-line argument
+`-max-lines`) and when it fills up, the oldest messages will be overwritten.
+
+To store the logs somewhere more permanent, for example a disk or a remote
+network service, a service should be added to the yaml which connects to
+`memlogd` and streams the logs. The example program `logread` in the `memlogd`
+package demonstrates how to do this.
+
+### Message format
+
+The format used to read logs is similar to [kmsg](https://www.kernel.org/doc/Documentation/ABI/testing/dev-kmsg):
+```
+<timestamp>,<origin>;<body>
+```
+where `<timestamp>` is an RFC3339-formatted timestamp, `<origin>` is the name of the
+original sending process and `<body>` is the output.
+
+### Usage examples
+```
+/ # logread -f
+2018-07-05T13:22:32Z,memlogd;memlogd started
+2018-07-05T13:22:32Z,onboot.001-dhcpcd.out;eth0: waiting for carrier
+2018-07-05T13:22:32Z,onboot.001-dhcpcd.err;eth0: could not detect a useable init
+ system
+2018-07-05T13:22:32Z,onboot.001-dhcpcd.out;eth0: carrier acquired
+2018-07-05T13:22:32Z,onboot.001-dhcpcd.out;DUID 00:01:00:01:22:d0:d8:18:02:50:00:00:00:02
+2018-07-05T13:22:32Z,onboot.001-dhcpcd.out;eth0: IAID 00:00:00:02
+2018-07-05T13:22:32Z,onboot.001-dhcpcd.out;eth0: adding address fe80::d33a:3936:
+2ee4:5c8c
+2018-07-05T13:22:32Z,onboot.001-dhcpcd.out;eth0: soliciting an IPv6 router
+2018-07-05T13:22:32Z,onboot.001-dhcpcd.out;eth0: soliciting a DHCP lease
+2018-07-05T13:22:32Z,onboot.001-dhcpcd.out;eth0: offered 192.168.65.4 from 192.1
+68.65.1 `vpnkit'
+2018-07-05T13:22:32Z,onboot.001-dhcpcd.out;eth0: leased 192.168.65.4 for 7200 se
+conds
+2018-07-05T13:22:32Z,onboot.001-dhcpcd.out;eth0: adding route to 192.168.65.0/24
+2018-07-05T13:22:32Z,onboot.001-dhcpcd.out;eth0: adding default route via 192.16
+8.65.1
+2018-07-05T13:22:32Z,onboot.001-dhcpcd.out;exiting due to oneshot
+2018-07-05T13:22:32Z,onboot.001-dhcpcd.out;dhcpcd exited
+^C
+```
+
+Current issues and limitations:
+
+- No docker logger plugin support yet - it could be nice to add support to
+  memlogd, so the docker container logs would also be gathered in one place
+- No syslog compatibility at the moment and `/dev/log` doesn’t exist. This
+  socket could be created to keep syslog compatibility, e.g. by using
+  https://github.com/mcuadros/go-syslog. Processes that require syslog should
+  then be able to log directly to memlogd.
```

Comment:
Do we have any processes/services in `pkg/` (or elsewehre we care about) which might want syslog?

Reply:
I don't know of any myself-- everything I've seen is ok with `stdout`/`stderr` logging. I guess the main advantage of `syslog` is that it's a bit more structured, since the client gets to choose a facility and level, where here with `stdout` and `stderr` it's all just lines.

Reply:
> I don't know of any myself

Cool. What I was really meaning was "if we know of any we should list them", if there are none that's even better ;-)